### ISSUE TITLE
Invalid lifetime elision hint

### ISSUE THREAD

**Issue by krojew** [rust-lang/rust-clippy#6622]
Lint name: needless_lifetimes

I tried this code:

```rust
fn items<'a>(&'a mut self, limit: usize) -> (Vec<&'a Item>, usize);
```

I expected to see this happen: no warning, since rustc requires explicit lifetime here

Instead, this happened: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)

### Meta

- `cargo clippy -V`: clippy 0.1.51 (202720bf 2021-01-21)
- `rustc -Vv`:
  ```
  rustc 1.51.0-nightly (202720bf4 2021-01-21)
  binary: rustc
  commit-hash: 202720bf483088dbdb343f78c0aa77067fdd8156
  commit-date: 2021-01-21
  host: x86_64-pc-windows-msvc
  release: 1.51.0-nightly
  LLVM version: 11.0.1
  ```

**Comment by camsteffen** [comment]
I was able to elide lifetimes for your example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=943b12504e32c3f5e2d8c22b9bc5a183)). Let me know if I'm missing something.

**Comment by krojew** [comment]
@camsteffen turns out it only occurs when building with `#[automock]` from the https://github.com/asomers/mockall crate.

**Comment by camsteffen** [comment]
Can you provide a minimal reproducible example?

**Comment by camsteffen** [comment]
Thanks! This seems to be an issue with `mockall` since it does not work with elided lifetimes. The github project has a couple reported issues related to returning references. I suggest adding `#[allow(clippy::needless_lifetimes)]` after `#[automock]`.
